### PR TITLE
feat: add license compliance check tool

### PR DESCRIPTION
**Pull request by pbredenberg**

### CODE REVIEW

File: scripts/check-license-compliance.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+   return new Promise((resolve, reject) => {
+      init(options, (err, packages) => {
+         if (err) {
+            reject(err);
+         }
+
+         resolve(packages);
+      });
+   });
+};
```
Comment (onebytegone) [inline scripts/check-license-compliance.js:30]: Would something like this work?

```js
const { init } = require('license-checker'),
      util = require('util'),
      initLicenseChecker = util.promisify(init);
```

File: .license-checker.cjs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+module.exports = {
+
+   allowList: [
```
Comment (onebytegone) [inline .license-checker.cjs:5]: What about this?
```suggestion
   permittedLicenses: [
```

File: scripts/check-license-compliance.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+   'ISC',
+   'Apache-2.0',
+   '(MIT OR Apache-2.0)',
+   'Public Domain',
+   'CC-BY-3.0',
+];
+
+(async () => {
+   const projectDirectoryPath = await packageDirectory(),
+         projectConfigPath = `${projectDirectoryPath}/.license-checker.cjs`,
```
Comment (onebytegone) [inline scripts/check-license-compliance.js:21]: Any particular reason this is .cjs and not just .js?

File: scripts/check-license-compliance.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            module: moduleName,
+            license: licenses,
+            ...moduleInfoReport[moduleName],
+         };
+      });
+
+   console.info(`This project has ${Object.values(licenseRecord).length} licenses.`);
+
+   unsupportedLicenses = Object.values(licenseRecord).filter((record) => {
+      return !ALLOWED_LICENSES.includes(record.license) && !allowList.includes(record.license);
```
Comment (onebytegone) [inline scripts/check-license-compliance.js:86]: If a project supplies a list of permitted licenses, should we keep checking the default license list?

File: scripts/check-license-compliance.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+         // normalize licenses since license-checker returns either a string or array
+         // of the license information
+         if (Array.isArray(licenses)) {
+            moduleInfoReport[moduleName].licenses.forEach((license) => {
+               licenseRecord[moduleName] = {
+                  module: moduleName,
+                  license: license,
+                  ...moduleInfoReport[moduleName],
+               };
+            });
```
Comment (onebytegone) [inline scripts/check-license-compliance.js:71]: This only keeps the last license. Would it be better to have a list of all the licenses and see if any are in the list of permitted licenses?

File: scripts/check-license-compliance.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      process.exitCode = 1;
+      return;
+   }
+
+   Object
+      .keys(moduleInfoReport)
+      .forEach((moduleName) => {
+         const licenses = moduleInfoReport[moduleName].licenses;
+
+         if (!licenses) {
```
Comment (onebytegone) [inline scripts/check-license-compliance.js:58]: If there are no licenses, should we assume that the package is `UNLICENSED` (i.e. the package cannot be used by anyone but the copyright holder)?

File: scripts/check-license-compliance.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      });
+   } catch(e) {
+      console.error('Could not initialize license-checker', e);
+      process.exitCode = 1;
+      return;
+   }
+
+   Object
+      .keys(moduleInfoReport)
+      .forEach((moduleName) => {
```
Comment (onebytegone) [inline scripts/check-license-compliance.js:55]: We'll also need a way to note what packages are permitted regardless of their license. Perhaps a config option like `permittedPackages`?

**Review by onebytegone — CHANGES_REQUESTED** [review]

File: scripts/check-license-compliance.js (hunk 4 above)
Comment (pbredenberg) [reply scripts/check-license-compliance.js:86]: I was imaging the licenses passed in would be _in addition_ to default set declared inside the tool.

File: scripts/check-license-compliance.js (hunk 6 above)
Comment (pbredenberg) [reply scripts/check-license-compliance.js:58]: Sure, good point!

File: scripts/check-license-compliance.js (hunk 5 above)
Comment (pbredenberg) [reply scripts/check-license-compliance.js:71]: You're right! I neglected to revisit this block of code. Here's the situation: in the many projects I've tested it in, I've never seen `const licenses = moduleInfoReport[moduleName].licenses;` evaluate to an array.

That's probably because [the library is handling an edge case](https://github.com/davglass/license-checker/blob/master/lib/index.js#L124) due to [what the NPM docs say here](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#license) regarding license data which they refer to as invalid.

In short, I don't think this block is actually needed from looking at the source linked above.

File: scripts/check-license-compliance.js (hunk 7 above)
Comment (pbredenberg) [reply scripts/check-license-compliance.js:55]: @onebytegone this hath been added, my liege

File: scripts/check-license-compliance.js (hunk 6 above)
Comment (pbredenberg) [reply scripts/check-license-compliance.js]: Did it

File: scripts/check-license-compliance.js (hunk 5 above)
Comment (pbredenberg) [reply scripts/check-license-compliance.js]: I'm now simply taking whatever value it sends. Let me know if you think we should still support this case.

File: .license-checker.cjs (hunk 2 above)
Comment (pbredenberg) [reply .license-checker.cjs]: Did it

File: scripts/check-license-compliance.js (hunk 3 above)
Comment (pbredenberg) [reply scripts/check-license-compliance.js]: Fixed
